src/t_zset.c

```diff
@@ -3819,71 +3845,108 @@ void genericZpopCommand(client *c, robj **keyv, int keyc, int where, int emitkey
             signalModifiedKey(c,c->db,key);
         }
 
-        if (use_nested_array) {
+        if (use_nested_array || use_nested_array_with_key) {
             addReplyArrayLen(c,2);
         }
         addReplyBulkCBuffer(c,ele,sdslen(ele));
         addReplyDouble(c,score);
         sdsfree(ele);
         ++result_count;
+    } while(--rangelen);
 
-        /* Remove the key, if indeed needed. */
-        if (zsetLength(zobj) == 0) {
-            dbDelete(c->db,key);
-            notifyKeyspaceEvent(NOTIFY_GENERIC,"del",key,c->db->id);
-            break;
-        }
-    } while(--count);
+    /* Remove the key, if indeed needed. */
+    if (zsetLength(zobj) == 0) {
+        if (deleted) *deleted = 1;
 
-    if (!use_nested_array) {
-        result_count *= 2;
+        dbDelete(c->db,key);
+        notifyKeyspaceEvent(NOTIFY_GENERIC,"del",key,c->db->id);
     }
-    setDeferredArrayLen(c,arraylen_ptr,result_count + (emitkey != 0));
 }
 
-/* ZPOPMIN key [<count>] */
-void zpopminCommand(client *c) {
+/* ZPOPMIN/ZPOPMAX key [<count>] */
+void zpopMinMaxCommand(client *c, int where) {
     if (c->argc > 3) {
         addReplyErrorObject(c,shared.syntaxerr);
         return;
     }
-    genericZpopCommand(c,&c->argv[1],1,ZSET_MIN,0,
-        c->argc == 3 ? c->argv[2] : NULL);
+
+    long count = 0; /* 0 for plain single pop. */
+    if (c->argc == 3) {
+        if (getLongFromObjectOrReply(c, c->argv[2], &count, NULL) != C_OK)
+            return;
+
+        if (count <= 0) {
+            addReply(c, shared.emptyarray);
+            return;
+        }
+    }
+
+    genericZpopCommand(c, &c->argv[1], 1, where, 0,
+                       count, 0, NULL);
+}
+
+/* ZPOPMIN key [<count>] */
+void zpopminCommand(client *c) {
+    zpopMinMaxCommand(c, ZSET_MIN);
 }
 
 /* ZMAXPOP key [<count>] */
 void zpopmaxCommand(client *c) {
-    if (c->argc > 3) {
-        addReplyErrorObject(c,shared.syntaxerr);
-        return;
-    }
-    genericZpopCommand(c,&c->argv[1],1,ZSET_MAX,0,
-        c->argc == 3 ? c->argv[2] : NULL);
+    zpopMinMaxCommand(c, ZSET_MAX);
 }
 
-/* BZPOPMIN / BZPOPMAX actual implementation. */
-void blockingGenericZpopCommand(client *c, int where) {
+/* BZPOPMIN, BZPOPMAX, BZMPOP actual implementation.
+ *
+ * 'numkeys' is the number of keys.
+ * 'timeout_idx' parameter position of block timeout.
+ * 'where' ZSET_MIN or ZSET_MAX.
+ * 'count' is the number of elements requested to pop, or 0 for plain single pop.
+ * 'use_nested_array_with_key' todo need a comment and a new name */
```

Comment:
i'd go with `use_nested_array` like in the other function.
it has a similar meaning, and just that in this case the emitkey is always true.
this way it's also trivial to forward it to the to the other function like we do below.